src/System Application/Test/Performance Profiler/src/ScheduledPerfProfilingTest.Codeunit.al

```diff
@@ -0,0 +1,217 @@
+// ------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// ------------------------------------------------------------------------------------------------
+
+namespace System.Test.Tooling;
+
+using System.PerformanceProfile;
+using System.Security.AccessControl;
+using System.TestLibraries.Utilities;
+using System.Tooling;
+
+codeunit 135019 "Scheduled Perf. Profiling Test"
+{
+    Subtype = Test;
+
+    var
+        Any: Codeunit Any;
+        Assert: Codeunit "Library Assert";
+        ScheduledPerfProfiler: Codeunit "Scheduled Perf. Profiler";
+        ProfileStartingDateLessThenEndingDateErr: Label 'The performance profile starting date must be set before the ending date.';
+        ProfileHasAlreadyBeenScheduledErr: Label 'Only one performance profile session can be scheduled for a given activity type for a given user for a given period.';
+
+    [Test]
+    procedure TestInitializedData()
+    var
+        TempPerformanceProfileScheduler: Record "Performance Profile Scheduler" temporary;
+        ActivityType: Enum "Activity Type";
+    begin
+        // [WHEN] The initial data shown on the "Perf. Profiler Schedules Card" card page is set up
+        ScheduledPerfProfiler.InitializeFields(TempPerformanceProfileScheduler, ActivityType);
+
+        // [THEN] Expected initalization happens
+        Assert.AreEqual(ActivityType, ActivityType::WebClient, 'Expected to be initialized to web client');
+        Assert.IsTrue(TempPerformanceProfileScheduler."Profile Creation Threshold" = 500, 'The default profile creation threshold is 500 ms.');
+        Assert.AreEqual(TempPerformanceProfileScheduler.Frequency, TempPerformanceProfileScheduler.Frequency::"100", 'The default frequency should be 100 ms.');
+        Assert.IsTrue(TempPerformanceProfileScheduler.Enabled, 'The scheduled sampling profile record should be enabled.');
+        Assert.IsFalse(IsNullGuid(TempPerformanceProfileScheduler."Schedule ID"), 'The scheduled sampling profile record should have been created a non zero guid.');
+        Assert.AreEqual(TempPerformanceProfileScheduler."User ID", UserSecurityId(), 'The scheduled sampling profile record should have been initialized with the user associated to the session.');
+    end;
+
+    [Test]
+    procedure TestMapRecordToActivityType()
+    var
+        TempPerformanceProfileScheduler: Record "Performance Profile Scheduler" temporary;
+        ExpectedActivityTypeMsg: Label 'Expected %1 actvity type. Actual type %2:', Locked = true;
+        ActivityType: Enum "Activity Type";
+    begin
+        // [WHEN] a web client session type is used
```

Comment:
In all tests you start with [WHEN] where it makes more sense to [GIVEN].

Please go to: https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-extension-advanced-example-test

Read more regards SCENARIO Tag and GIVEN-WHEN-THEN Tags

![image](https://github.com/microsoft/BCApps/assets/59221826/684fbdca-2ff5-450a-a8b8-dbe9230b1456)

Structure of all tests should be adjusted

Reply:
@BazookaMusic After the most recent commit, still adjustments needed. It's better that there are GIVEN tags now, but WHEN is missing. Also, be aware that it could be multiple GIVEN tags.

Reply:
@kalberes Answer is wrong in many ways. I don't agree with this statement at all. 

**First**, the initial point is to structure it by tags, not to put all in [WHEN], and that's it, it's more correct than not to use any tags. Also, using [WHEN] where it was [GIVEN], it's just the wrong tag, if you don't care about tags better not use it, than just use it on wrong way. 
**Second**, a rough description is fine. You can read more about it on https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-extension-advanced-example-test 
Or take a look at other tests which are existing though app. It should not explain what code does, but explain what test does, what is given, what are you doing and what are you verifying.
**Third**, if all tests follow this guideline why wouldn't this one? Consistency. 
**Fourth**, it would be the same as saying you don't need documentation for anything, just look up in code, and yet Docs exist.

It's excuse for not wanting to follow simple guideline, which is consistent on all tests. Is it something that takes too much time? It can take 5 mins to adjust this test Codeunit. I am more for approach if you do it, than do it as it should be done, on right way. Don't throws just random tags in your test procedures and than give excuses for that.

Reply:
@JesperSchulz @nikolakukrika 
Could you guys take a look at this test structure? Do we just throw any tags around in test procedures?

Reply:
The tags were fixed yesterday before the discussion, so there's no longer "GIVEN" tags in place of [WHEN]. On the one hand, this is not the place to change this convention because all other tests follow it. 

However, to Kalman's point, this convention is not ideal for maintainability. Comments are not in the type system and tend to get outdated as code ages. They can even be misleading as code is moved around.

Ideally we could have a unit testing library that expresses this flow, however the lack of anonymous functions probably makes this hard at the moment.